### PR TITLE
Encrypt and Decrypt commands

### PR DESCRIPTION
**Pull request by fboucquez**
Fixes https://github.com/nemtech/symbol-bootstrap/issues/154

-   **A valid password is required by default**. Bootstrap will ask for the password, or you can provide one with `--password`.
-   Added `--noPassword` to bypass the password validation. Private keys in generated addresses.yml and preset.yml will be in plain text!
-   Added `encrypt` and `decrypt` commands.

@ravitej-aluru @Wayonb, if this PR goes in, you will need to update your scripts adding --noPassword

### CODE REVIEW

File: src/commands/decrypt.ts [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+import { existsSync } from 'fs';
+import { dirname } from 'path';
+import { LogType } from '../logger';
+import Logger from '../logger/Logger';
+import LoggerFactory from '../logger/LoggerFactory';
+import { BootstrapUtils, KnownError } from '../service';
+const logger: Logger = LoggerFactory.getLogger(LogType.System);
+
+export default class Decrypt extends Command {
+    static description = `It decrypts a yml file using the provided password. The source files would be a custom preset file, a preset.yml file or an addresses.yml.
```
Comment (segfaultxavi) [inline src/commands/decrypt.ts:27]:
```suggestion
    static description = `It decrypts a yml file using the provided password. The source file can be a custom preset file, a preset.yml file or an addresses.yml.
```

File: src/commands/decrypt.ts [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+`;
+
+    static examples = [
+        `$ symbol-bootstrap decrypt --password 1234 --source plain-custom-preset.yml --destination decrypted-custom-preset.yml`,
+    ];
+
+    static flags = {
+        help: BootstrapUtils.helpFlag,
+        source: flags.string({
+            description: `The source plain yml file to be decrypted. If this file is not decrypted, the command will raise an error.`,
```
Comment (segfaultxavi) [inline src/commands/decrypt.ts:48]:
```suggestion
            description: `The source encrypted yml file to be decrypted. If this file is not encrypted, the command will raise an error.`,
```

File: src/commands/decrypt.ts [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+            required: true,
+        }),
+        destination: flags.string({
+            description: `The destination where the decrypted file will be stored. The destination file must not exist.`,
```
Comment (segfaultxavi) [inline src/commands/decrypt.ts:52]:
```suggestion
            description: `The destination decrypted file to create. The destination file must not exist.`,
```

File: src/commands/decrypt.ts [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+            required: true,
+        }),
+        password: flags.string({
+            description: `The password used to decrypt the source file into the destination file. Keep this password in a secure place!`,
```
Comment (segfaultxavi) [inline src/commands/decrypt.ts:56]:
```suggestion
            description: `The password to use to decrypt the source file into the destination file. Keep this password in a secure place!`,
```

File: src/commands/decrypt.ts [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        }
+        if (existsSync(flags.destination)) {
+            throw new KnownError(`Destination file ${flags.destination} already exist!`);
+        }
+        BootstrapUtils.validatePassword(flags.password);
+        const data = await BootstrapUtils.loadYaml(flags.source, flags.password);
+        await BootstrapUtils.mkdir(dirname(flags.destination));
+        await BootstrapUtils.writeYaml(flags.destination, data, '');
+        logger.info(
+            `Decrypted file ${flags.destination} has been created! Private keys on this files are in plain text. Remember to remove the file!`,
```
Comment (segfaultxavi) [inline src/commands/decrypt.ts:76]:
```suggestion
            `Decrypted file ${flags.destination} has been created! Any private keys on this file are now in plain text. Remember to remove the file!`,
```

File: src/commands/encrypt.ts [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+import { dirname } from 'path';
+import { LogType } from '../logger';
+import Logger from '../logger/Logger';
+import LoggerFactory from '../logger/LoggerFactory';
+import { BootstrapUtils, KnownError } from '../service';
+import { CryptoUtils } from '../service/CryptoUtils';
+const logger: Logger = LoggerFactory.getLogger(LogType.System);
+
+export default class Encrypt extends Command {
+    static description = `It encrypts a yml file using the provided password. The source files would be a custom preset file, a preset.yml file or an addresses.yml.
```
Comment (segfaultxavi) [inline src/commands/encrypt.ts:28]:
```suggestion
    static description = `It encrypts a yml file using the provided password. The source file can be a custom preset file, a preset.yml file or an addresses.yml.
```

File: src/commands/encrypt.ts [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+The main use case of this command is encrypting custom presets files. If your custom preset contains private keys, it's highly recommended to encrypt it and use provide --password when starting or configuring the node with Bootstrap.
```
Comment (segfaultxavi) [inline src/commands/encrypt.ts:30]:
```suggestion
The main use case of this command is encrypting custom presets files. If your custom preset contains private keys, it's highly recommended to encrypt it and provide --password when starting or configuring the node with Bootstrap.
```

File: src/commands/encrypt.ts [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+`;
+
+    static examples = [
+        `$ symbol-bootstrap encrypt --password 1234 --source plain-custom-preset.yml --destination encrypted-custom-preset.yml`,
+    ];
+
+    static flags = {
+        help: BootstrapUtils.helpFlag,
+        source: flags.string({
+            description: `The source plain yml file to be encrypted. If this file is encrypted, the command will raise an error.`,
```
Comment (segfaultxavi) [inline src/commands/encrypt.ts:44]:
```suggestion
            description: `The source plain yml file to be encrypted. If this file is already encrypted, the command will raise an error.`,
```

File: src/commands/encrypt.ts [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+            required: true,
+        }),
+        destination: flags.string({
+            description: `The destination where the encrypted file will be stored. The destination file must not exist.`,
```
Comment (segfaultxavi) [inline src/commands/encrypt.ts:48]:
```suggestion
            description: `The destination encrypted file to create. The destination file must not exist.`,
```

File: src/commands/encrypt.ts [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+            required: true,
+        }),
+        password: flags.string({
+            description: `The password used to encrypt the source file into the destination file. Keep this password in a secure place!`,
```
Comment (segfaultxavi) [inline src/commands/encrypt.ts:52]:
```suggestion
            description: `The password to use to encrypt the source file into the destination file. Keep this password in a secure place!`,
```

File: src/commands/encrypt.ts [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+        if (!existsSync(flags.source)) {
+            throw new KnownError(`Source file ${flags.source} does not exist!`);
+        }
+        if (existsSync(flags.destination)) {
+            throw new KnownError(`Destination file ${flags.destination} already exist!`);
+        }
+        BootstrapUtils.validatePassword(flags.password);
+        const data = await BootstrapUtils.loadYaml(flags.source, false);
+        if (CryptoUtils.encryptedCount(data) > 0) {
+            throw new KnownError(`Source file ${flags.source} is encrypted already. Do you want to decrypt? Try the decrypt command.`);
```
Comment (segfaultxavi) [inline src/commands/encrypt.ts:70]:
```suggestion
            throw new KnownError(`Source file ${flags.source} is already encrypted. If you want to decrypt it use the decrypt command.`);
```

File: src/commands/decrypt.ts [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+    };
+
+    public async run(): Promise<void> {
+        const { flags } = this.parse(Decrypt);
+
+        if (!existsSync(flags.source)) {
+            throw new KnownError(`Source file ${flags.source} does not exist!`);
+        }
+        if (existsSync(flags.destination)) {
+            throw new KnownError(`Destination file ${flags.destination} already exist!`);
```
Comment (segfaultxavi) [inline src/commands/decrypt.ts:69]:
```suggestion
            throw new KnownError(`Destination file ${flags.destination} already exists!`);
```

File: src/commands/encrypt.ts [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+    };
+
+    public async run(): Promise<void> {
+        const { flags } = this.parse(Encrypt);
+
+        if (!existsSync(flags.source)) {
+            throw new KnownError(`Source file ${flags.source} does not exist!`);
+        }
+        if (existsSync(flags.destination)) {
+            throw new KnownError(`Destination file ${flags.destination} already exist!`);
```
Comment (segfaultxavi) [inline src/commands/encrypt.ts:65]:
```suggestion
            throw new KnownError(`Destination file ${flags.destination} already exists!`);
```

**Review by segfaultxavi — CHANGES_REQUESTED** [review]

File: src/service/CommandUtils.ts [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+        description: 'The target folder where the symbol-bootstrap network is generated',
+        default: BootstrapUtils.defaultTargetFolder,
+    });
+
+    public static passwordFlag = CommandUtils.getPasswordFlag(
+        `A password used to encrypt and decrypted custom presets, addresses.yml, and preset.yml files. When providing a password, private keys would be encrypted. Keep this password in a secure place!`,
+    );
+
+    public static noPasswordFlag = flags.boolean({
+        description: 'When provided, Bootstrap will not ask for a password',
```
Comment (segfaultxavi) [inline src/service/CommandUtils.ts:40]:
```suggestion
        description: 'When provided, Bootstrap will not use a password, so private keys will be stored in plain text. Use with caution.',
```

**Review by segfaultxavi — CHANGES_REQUESTED** [review]

File: docs/start.md [hunk 15]
```diff
@@ -56,6 +56,13 @@ OPTIONS
 
       The health check process handles 'repeat' and custom 'openPort' services.
 
+  --noPassword
+      When provided, Bootstrap will not use a password, so private keys will be stored in plain text. Use with caution.
+
+  --password=password
```
Comment (Wayonb) [inline docs/start.md:62]: Its a security risk to take the password on the command line.
if a user node gets hacked, the hacker can look through the shell history to find the user password.
would suggest add another parameter to give the user a more secure option.

      --password-stdin    Take the password from stdin

**Review by Wayonb — CHANGES_REQUESTED** [review]
Looks good.   One issue above which I think should be fixed.

File: docs/start.md (hunk 15 above)
Comment (fboucquez) [reply docs/start.md:62]: Yes. Try it out. If password is not provided, bootstrap will ask for it

Comment (fboucquez) [reply docs/start.md:62]: I've just tried pipe | commands like the example from docker and it works really nice.
````
echo "$MY_ENV_VAR_PASSWORD" | symbol-bootstrap start -p testnet -a dual
````
````
echo "$MY_ENV_VAR_PASSWORD" | symbol-bootstrap decrypt --source target/addresses.yml --destination plain-addresses.yml
````
Added examples

Comment (Wayonb) [reply docs/start.md:62]: cool.  Sounds good.

**Review by Wayonb — APPROVED** [review]

File: docs/start.md (hunk 15 above)
Comment (fboucquez) [reply docs/start.md:62]: I'll merge after tomorrow's supernode release.

File: src/service/CommandUtils.ts [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+    public static helpFlag = flags.help({ char: 'h', description: 'It shows the help of this command.' });
+
+    public static targetFlag = flags.string({
+        char: 't',
+        description: 'The target folder where the symbol-bootstrap network is generated',
+        default: BootstrapUtils.defaultTargetFolder,
+    });
+
+    public static passwordFlag = CommandUtils.getPasswordFlag(
+        `A password used to encrypt and decrypted custom presets, addresses.yml, and preset.yml files. When providing a password, private keys would be encrypted. Keep this password in a secure place!`,
```
Comment (segfaultxavi) [inline src/service/CommandUtils.ts:36]:
```suggestion
        `A password used to encrypt and decrypt private keys in preset files like addresses.yml and preset.yml. Password can be provided in the command line (--password=XXXX) or requested in a prompt (--password).`,
```

File: src/commands/decrypt.ts [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+        source: flags.string({
+            description: `The source encrypted yml file to be decrypted.`,
+            required: true,
+        }),
+        destination: flags.string({
+            description: `The destination decrypted file to create. The destination file must not exist.`,
+            required: true,
+        }),
+        password: CommandUtils.getPasswordFlag(
+            `The password to use to decrypt the source file into the destination file. Keep this password in a secure place!`,
```
Comment (segfaultxavi) [inline src/commands/decrypt.ts:70]:
```suggestion
            `The password to use to decrypt the source file into the destination file. Password can be provided in the command line (--password=XXXX) or requested in a prompt (--password).`,
```

File: src/commands/encrypt.ts [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+        source: flags.string({
+            description: `The source plain yml file to be encrypted. If this file is encrypted, the command will raise an error.`,
+            required: true,
+        }),
+        destination: flags.string({
+            description: `The destination encrypted file to create. The destination file must not exist.`,
+            required: true,
+        }),
+        password: CommandUtils.getPasswordFlag(
+            `The password to use to encrypt the source file into the destination file. Keep this password in a secure place!`,
```
Comment (segfaultxavi) [inline src/commands/encrypt.ts:60]:
```suggestion
            `The password to use to encrypt the source file into the destination file. Password can be provided in the command line (--password=XXXX) or requested in a prompt (--password).`,
```

**Review by segfaultxavi — CHANGES_REQUESTED** [review]
It's important that we clarify the two modes of the --password flag.

File: src/commands/decrypt.ts (hunk 17 above)
Comment (fboucquez) [reply src/commands/decrypt.ts:70]: @segfaultxavi What about:

> The password to use to decrypt the source file into the destination file. Bootstrap prompts for a password by default, can be provided in the command line (--password=XXXX) or disabled in the command line (--noPassword).

File: src/commands/encrypt.ts (hunk 18 above)
Comment (fboucquez) [reply src/commands/encrypt.ts:60]: What about. Prompt is use by default. 

>A password used to encrypt and decrypt private keys in preset files like addresses.yml and preset.yml. Bootstrap prompts for a password by default, can be provided in the command line (--password=XXXX) or disabled in the command line (--noPassword).

File: src/service/CommandUtils.ts (hunk 16 above)
Comment (fboucquez) [reply src/service/CommandUtils.ts:36]: What about. 

> A password used to encrypt and decrypt private keys in preset files like addresses.yml and preset.yml. Bootstrap prompts for a password by default, can be provided in the command line (--password=XXXX) or disabled in the command line (--noPassword).

Prompt is used by default.

File: src/commands/encrypt.ts (hunk 18 above)
Comment (segfaultxavi) [reply src/commands/encrypt.ts]: Sounds good! 👍

**Pull request merged**